### PR TITLE
Update Azure SDK semantic conventions and guidelines

### PR DESCRIPTION
**Pull request by lmolkova**
Fixes https://github.com/Azure/azure-sdk-for-net/issues/39785

### CODE REVIEW

File: docs/tracing/distributed-tracing-conventions.md [hunk 1]
```diff
@@ -6,25 +6,286 @@ folder: general
 sidebar: general_sidebar
 ---
 
-Conventions are the contract between Azure SDK and tracing providers such as Azure Monitor, Jaeger and others.  They describe and standardize attributes, events and relationships for common span types: HTTP, DB, messaging and others.  Observability vendors use conventions to build visualizations and may be very sensitive to them.  Custom Azure SDK conventions are described in [tracing-conventions.yml](./distributed-tracing-conventions.yml) and include:
-
-- Custom Azure SDK attributes
-- Azure SDK HTTP conventions compatible with OpenTelemetry
-- Custom messaging and DB conventions. They don't follow OpenTelemetry convention and used until OpenTelemetry conventions evolve and stabilize
+Conventions are the contract between Azure SDK and tracing providers such as Azure Monitor, Jaeger and others.  They describe and standardize attributes, events and relationships for common span types: HTTP, DB, messaging and others.  Observability vendors use conventions to build visualizations and may be very sensitive to them.  Custom Azure SDK conventions are described in [Azure SDK semantic conventions](#azure-sdk-semantic-conventions) section below.
```
Comment (JoshLove-msft) [inline docs/tracing/distributed-tracing-conventions.md:9]:
```suggestion
Conventions are the contract between Azure SDK and tracing providers such as Azure Monitor, Jaeger, and others.  They describe and standardize attributes, events and relationships for common span types: HTTP, DB, messaging, and others.  Observability vendors use conventions to build visualizations and may be very sensitive to them.  Custom Azure SDK conventions are described in [Azure SDK semantic conventions](#azure-sdk-semantic-conventions) section below.
```

File: docs/tracing/distributed-tracing-conventions.md [hunk 2]
```diff
@@ -6,25 +6,286 @@ folder: general
...
 
+{% include requirement/SHOULD id="general-tracing-convention-set-library" %} set Azure Client Library name and version [Schema URL](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#get-a-tracer) when [creating OpenTelemetry tracer](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#get-a-tracer).
 
 {% include requirement/SHOULD id="general-tracing-convention-new-otel" %} contribute new conventions (or patch existing ones) to OpenTelemetry when there is no suitable one or some scenarios are missing.
 
-{% include requirement/MAY id="general-tracing-convention-add-attributes" %} extend list of attributes on top of OpenTelemetry contentions with Azure-specific ones.
+# Azure SDK semantic conventions
+
+<!--TODO: move to https://github.com/open-telemetry/semantic-conventions repo-->
```
Comment (JoshLove-msft) [inline docs/tracing/distributed-tracing-conventions.md:27]: Should this be removed?

File: docs/tracing/distributed-tracing-conventions.md [hunk 3]
```diff
@@ -6,25 +6,286 @@ folder: general
...
+
+# Semantic conventions for Azure SDK spans
```
Comment (JoshLove-msft) [inline docs/tracing/distributed-tracing-conventions.md:29]: This is at the same heading level as L25

File: docs/tracing/distributed-tracing-conventions.md [hunk 4]
```diff
@@ -6,25 +6,286 @@ folder: general
...
+
+**Status**: [Mixed][DocumentStatus]
+
+This document describes tracing semantic conventions adopted by Azure SDK. Instrumentations live in Azure SDK repos and are shipped along with Azure SDK artifacts.
```
Comment (JoshLove-msft) [inline docs/tracing/distributed-tracing-conventions.md:33]: Can you clarify what is meant by "Instrumentations" here?

File: docs/tracing/distributed-tracing-conventions.md [hunk 5]
```diff
@@ -6,25 +6,286 @@ folder: general
...
+
+- [Java](https://github.com/Azure/azure-sdk-for-java/tree/main/sdk/core/azure-core-tracing-opentelemetry)
+- [JavaScript](https://github.com/Azure/azure-sdk-for-js/tree/main/sdk/core/core-tracing)
+- [Python](https://github.com/Azure/azure-sdk-for-python/tree/main/sdk/core/azure-core-tracing-opentelemetry)
+- [.NET](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/samples/Diagnostics.md#activitysource-support)
+- [Go](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/tracing/azotel)
+- [C++](https://github.com/Azure/azure-sdk-for-cpp/tree/main/sdk/core/azure-core-tracing-opentelemetry)
+
+Azure SDK produces spans for public API calls and nested HTTP client spans. AMQP transport-level calls are not traced.
```
Comment (JoshLove-msft) [inline docs/tracing/distributed-tracing-conventions.md:42]:
```suggestion
The Azure SDK produces spans for public API calls and nested HTTP client spans. AMQP transport-level calls are not traced.
```

File: docs/tracing/distributed-tracing-conventions.md [hunk 6]
```diff
@@ -6,25 +6,286 @@ folder: general
...
+
+## Versioning
+
+Azure SDKs follow OpenTelemetry semantic conventions when applicable, but adopt new versions of conventions at their own pace. Telemetry consumers MAY use [SchemaUrl](https://opentelemetry.io/docs/specs/otel/schemas/#schema-url) to detect which version of semantic conventions are emitted by Azure SDKs.
```
Comment (JoshLove-msft) [inline docs/tracing/distributed-tracing-conventions.md:46]: I think we either say "Azure SDK" when referring to all libraries, or "Azure client libraries"

File: docs/tracing/distributed-tracing-conventions.md [hunk 7]
```diff
@@ -6,25 +6,286 @@ folder: general
...
+
+## Public API calls
+
+**Status**: [Stable][DocumentStatus]
+
+Azure SDK SHOULD create a span for each call to service methods, that is, public APIs that involve communication with Azure services.
```
Comment (JoshLove-msft) [inline docs/tracing/distributed-tracing-conventions.md:52]:
```suggestion
Azure SDK SHOULD create a span for each call to public APIs that involve communication with Azure services.
```

File: docs/tracing/distributed-tracing-conventions.md [hunk 8]
```diff
@@ -6,25 +6,286 @@ folder: general
...
+
+- Spans representing public API calls SHOULD have names following `client.method` pattern and are language-specific. In case OpenTelemetry defines semantic convention for span name (for example, in messaging or database conventions), standard OpenTelemetry name SHOULD be used instead.
+- For HTTP-based SDKs, public API spans SHOULD have `INTERNAL` kind.
```
Comment (JoshLove-msft) [inline docs/tracing/distributed-tracing-conventions.md:55]:
```suggestion
- For HTTP-based SDKs, public API spans SHOULD have kind of `INTERNAL`.
```

Comment (JoshLove-msft) [reply docs/tracing/distributed-tracing-conventions.md:55]: Could we link to definition of kind?

File: docs/tracing/distributed-tracing-conventions.md [hunk 9]
```diff
@@ -6,25 +6,286 @@ folder: general
...
+**[1]:** This SHOULD be set as an instrumentation scope attribute when creating a `Tracer` as long as OpenTelemetry in a given language allows to do so.
+
+**[2]:** if and only if OpenTelemetry in a given language doesn't provide a standard way to set `schema_url` (.NET)
+
+**[3]:** If the request fails with an error before response status code was sent or received,
+`error.type` SHOULD be set to exception type (its fully-qualified class name, if applicable)
+or a component-specific low cardinality error identifier.
+
+If response status code was sent or received and status indicates an error according to [HTTP span status definition](/docs/http/http-spans.md),
+`error.type` SHOULD be set to the status code number (represented as a string), an exception type (if thrown) or a component-specific error identifier.
```
Comment (JoshLove-msft) [inline docs/tracing/distributed-tracing-conventions.md:199]: Worth mentioning the fallback value here?

File: docs/tracing/distributed-tracing-conventions.md [hunk 10]
```diff
@@ -6,25 +6,286 @@ folder: general
...
+* [`url.full`](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/attributes-registry/url.md)
+<!-- endsemconv -->
+
+Instrumentation supports [W3C Trace context](https://w3c.github.io/trace-context/) propagation and Azure legacy correlation protocols. Propagator configuration is not supported.
+
+## Messaging SDKs
+
+**Status**: [Experimental][DocumentStatus]
+
+Messaging span semantics apply to Azure Event Hubs and Service Bus SDKs and follow [OpenTelemetry Messaging spans conventions v1.22.0](https://github.com/open-telemetry/semantic-conventions/blob/v1.22.0/docs/messaging/messaging-spans.md).
```
Comment (JoshLove-msft) [inline docs/tracing/distributed-tracing-conventions.md:239]: Have we given thought to whether Event Grid should adopt the messaging semantic conventions? The V2 service allows for sending and receiving events in a similar way to Service Bus.

**Review by JoshLove-msft — DISMISSED** [review]
The language agnostic and .NET bits LGTM.

File: docs/tracing/distributed-tracing-conventions.md (hunk 2 above)
Comment (lmolkova) [reply docs/tracing/distributed-tracing-conventions.md:27]: I still want to move them there and this line is invisible for readers, but gives me extra motivation any time I see it :)

File: docs/tracing/distributed-tracing-conventions.md (hunk 9 above)
Comment (lmolkova) [reply docs/tracing/distributed-tracing-conventions.md:199]: the exception type is the fallback, rephrased

File: docs/tracing/distributed-tracing-conventions.md (hunk 10 above)
Comment (lmolkova) [reply docs/tracing/distributed-tracing-conventions.md:239]: great point, added https://github.com/Azure/azure-sdk/issues/6954 to track it

File: docs/tracing/distributed-tracing-conventions.md (hunk 3 above)
Comment (lmolkova) [reply docs/tracing/distributed-tracing-conventions.md:29]: not quite, we'll have some metrics in the future and wanted to emphasize it's tracing ones. Added a sentence to clarify.

**Review by jsquire — APPROVED** [review]

File: docs/tracing/distributed-tracing-conventions.md (hunk 9 above)
Comment (JoshLove-msft) [reply docs/tracing/distributed-tracing-conventions.md]: I thought _OTHER was the last fallback.

**Pull request merged**